sw/host/tests/usbdev/usbdev_stream/stream_test.cc

```diff
@@ -0,0 +1,273 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// USB streaming data test
+//
+// Linux host-side application that receives a stream of LFSR-generated data
+// from the USB device, checks the received bytestream and then XORs it with a
+// host-side LFSR-generated byte stream to transmit back to the device.
+//
+// By default the streaming test expects a number of USB serial connections to
+// the target device, one port per endpoint:
+//
+//   /dev/ttyUSB0 - supplies and receives LFSR-generated byte stream for one/
+//                  the only endpoint
+//   /dev/ttyUSB1 - a secondary stream
+//   /dev/ttyUSB..
+//
+// Note that the mapping from device endpoints to USB port number is not
+// guaranteed, and  when multiple streams are used, it is _not_ necessarily the
+// case that ascending streams/endpoints in usbdev_stream_test are mapped to
+// a contiguous range of ascending ttyUSBi port names.
+//
+// Either or both of the initial input port and the initial output port may be
+// overridden using command line parameters.
+//
+// Usage:
+//   stream [-n<streams>][-v<bool>][-c<bool>][-r<bool>][-s<bool>]
+//          [<input port>[ <output port>]]
+//
+//   -c   check any retrieved data against expectations
+//   -n<> Number of concurrent byte streams to test
+//   -r   retrieve data from device
+//   -s   send data to device
+//   -v   verbose reporting
+//
+// Build:
+//   eg. g++ -Wall -Werror -o stream_test
+//           stream_test.cc usbdev_stream.cc usbdev_utils.cc
+
+#include "stream_test.h"
+
+#include <cassert>
+#include <cctype>
+#include <cerrno>
+#include <cinttypes>
+#include <cstdbool>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <ctime>
+#include <sys/time.h>
+
+#include "usbdev_stream.h"
+#include "usbdev_utils.h"
+
+// Test properties
+#ifndef BYTES_TO_TRANSFER
+// This is about the amount that we can transfer within a 1 hour 'eternal' test
+//#define BYTES_TO_TRANSFER (0x60U << 20)
+// This takes about 80s presently with transfer of sending of undefined data
+// and no checking of the received data, but ca. 213s with LFSR generation and
+// checking
```

Comment:
I assume the difference is due to the performance bottleneck on the device side?

Reply:
Yes, the Ibex operates at just 10MHz in current CW310 builds, so even pretty minimal LFSR-related code, plus the inefficiencies in our memcpy_to|from_mmio32 implementation make a big difference to the performance.